libs/community/langchain_community/vectorstores/cratedb/base.py

```diff
@@ -0,0 +1,474 @@
+            data=kwargs,
+            key="connection_string",
+            env_key="CRATEDB_CONNECTION_STRING",
+        )
+
+        if not connection_string:
+            raise ValueError(
+                "Database connection string is required."
+                "Either pass it as a parameter, or set the "
+                "CRATEDB_CONNECTION_STRING environment variable."
+            )
+
+        return connection_string
+
+    @classmethod
+    def connection_string_from_db_params(
+        cls,
+        driver: str,
+        host: str,
+        port: int,
+        database: str,
+        user: str,
+        password: str,
+    ) -> str:
+        """Return connection string from database parameters."""
+        return str(
+            sqlalchemy.URL.create(
+                drivername=driver,
+                host=host,
+                port=port,
+                username=user,
+                password=password,
+                query={"schema": database},
+            )
+        )
+
+    def _select_relevance_score_fn(self) -> Callable[[float], float]:
+        """
+        The 'correct' relevance function
+        may differ depending on a few things, including:
+        - the distance / similarity metric used by the VectorStore
+        - the scale of your embeddings (OpenAI's are unit normed. Many others are not!)
+        - embedding dimensionality
+        - etc.
+        """
+        if self.override_relevance_score_fn is not None:
+            return self.override_relevance_score_fn
+
+        # Default strategy is to rely on distance strategy provided
+        # in vectorstore constructor
+        if self._distance_strategy == DistanceStrategy.COSINE:
+            return self._cosine_relevance_score_fn
+        elif self._distance_strategy == DistanceStrategy.EUCLIDEAN:
+            return self._euclidean_relevance_score_fn
+        elif self._distance_strategy == DistanceStrategy.MAX_INNER_PRODUCT:
+            return self._max_inner_product_relevance_score_fn
+        else:
+            raise ValueError(
+                "No supported normalization function for distance_strategy of "
+                f"{self._distance_strategy}. Consider providing relevance_score_fn to "
+                "CrateDBVectorStore constructor."
+            )
+
+    @staticmethod
+    def _euclidean_relevance_score_fn(similarity: float) -> float:
+        """Return a similarity score on a scale [0, 1]."""
+        # The 'correct' relevance function
+        # may differ depending on a few things, including:
+        # - the distance / similarity metric used by the VectorStore
+        # - the scale of your embeddings (OpenAI's are unit normed. Many
+        #  others are not!)
+        # - embedding dimensionality
+        # - etc.
+        # This function converts the Euclidean norm of normalized embeddings
+        # (0 is most similar, sqrt(2) most dissimilar)
+        # to a similarity function (0 to 1)
+
+        # Original:
```

Comment:
Should we remove this and next line?

Reply:
Thanks. 4b9310e517 improves that section, removing the lines you flagged, and replacing them with a better description about CrateDB's implementation.